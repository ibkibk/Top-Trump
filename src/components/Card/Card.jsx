import React, { Component } from "react";
import styles from "./Card.module.scss";
import PropTypes from "prop-types";

// Class containing JSX & methods associated with the card component

class Card extends Component {
  // Function check the length of the card's description & either returns it whole or shortened
  getDescription = () => {
    const description = this.props.cardData.description;
    return description.length > 120
      ? description.substring(0, 120) + "..."
      : description;
  };

  // object containing any properties related to component state
  state = {
    isFaceUp: true
  };

  // setState allows us to update state indirectly
  handleClick = () => {
    this.setState({
      isFaceUp: !this.state.isFaceUp
    });
  };

  /*
    render function describes what the UI will look like using JSX.
  */

  render() {
    const rotateStyle = this.state.isFaceUp ? "" : styles.cardRotated;
    return (
      <div className={styles.scene}>
        <article
          onClick={this.handleClick}
          className={`${styles.card} ${rotateStyle}`}
        >
          <div className={styles.cardFront}>
            <h3>{this.props.cardData.name}</h3>
            <div className={styles.topSection}>
              <section className={styles.description}>
                {this.getDescription()}
              </section>
              <img
                className={styles.picture}
                alt={this.props.cardData.name + "'s face"}
                src={this.props.cardData.picture}
              />
            </div>
            <table className={styles.skills}>
              <tbody>
                <tr>
                  <th>Loudness</th>
                  <td>{this.props.cardData.skills.loudness}</td>
                </tr>
                <tr>
                  <th>Greed</th>
                  <td>{this.props.cardData.skills.greed}</td>
                </tr>
                <tr>
                  <th>Attitude</th>
                  <td>{this.props.cardData.skills.attitude}</td>
                </tr>
                <tr>
                  <th>Sass</th>
                  <td>{this.props.cardData.skills.sass}</td>
                </tr>
                <tr>
                  <th>MVP</th>
                  <td>{this.props.cardData.skills.mvp}</td>
                </tr>
              </tbody>
            </table>
          </div>
          <div className={styles.cardBack} />
        </article>
      </div>
    );
  }
}

// Using propTypes to define the data types for the 'skills' object

const skillsTypes = PropTypes.shape({
  loudness: PropTypes.number.isRequired,
  greed: PropTypes.number.isRequired,
  attitude: PropTypes.number.isRequired,
  sass: PropTypes.number.isRequired,
  mvp: PropTypes.number.isRequired
});

// Using propTypes to define the data types for
// the props object passed to the card component

Card.propTypes = {
  cardData: PropTypes.shape({
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    picture: PropTypes.string.isRequired,
    skills: skillsTypes
  })
};

export default Card;
