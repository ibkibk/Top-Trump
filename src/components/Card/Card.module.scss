/* playing cards ratio is 2.5:3.5 */

/* SASS variable setup */
$card-width: 265px;
$card-height: $card-width * 3.5 / 2.5;
$picture-width: 100px;
$picture-height: $picture-width * 3.5 / 2.5;

.scene {
  width: $card-width;
  min-width: $card-width;
  height: $card-height;
  perspective: 600px;
  margin: 10px;
}

.card {
  text-align: center;
  width: 100%;
  height: 100%;
  position: relative;
  transition: transform 1s;
  transform-style: preserve-3d;

  h3 {
    margin-bottom: 10px;
  }
}

%card {
  border: 8px solid black;
  background-size: cover;
  border-radius: 15px;
  background-repeat: no-repeat, repeat;
  position: absolute;
  height: calc(100% - 16px);
  width: calc(100% - 16px);
  backface-visibility: hidden;
}

.cardFront {
  @extend %card;
  background-image: url(../../static/images/card.png);
}

.cardBack {
  @extend %card;
  background-image: url(../../static/images/card-back.png);
  transform: rotateY(180deg);
}

.cardRotated {
  transform: rotateY(180deg);
}

.picture {
  width: $picture-width;
  height: $picture-height;
  border: 1px solid black;
}

.topSection {
  display: flex;
  justify-content: space-around;
}

.description {
  width: $picture-width;
  height: $picture-height;
  font-size: small;
}

.skills {
  margin: 5px auto 5px auto;
  width: 100%;
  border-spacing: 5px 5px;
  td,
  th {
    background-color: black;
    color: white;
    border: 1px solid white;
    padding: 2px 15px 2px 15px;
    text-align: right;
  }
  td {
    border-top-left-radius: 25px;
  }
  th {
    border-bottom-right-radius: 25px;
  }
}
