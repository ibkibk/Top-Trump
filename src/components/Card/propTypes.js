import PropTypes from "prop-types";

// Using propTypes to define the data types for the 'skills' object

export const skillsTypes = PropTypes.objectOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number]));

// Using propTypes to define the data types for
// the props object passed to the card component

export const cardTypes = {
  cardData: PropTypes.shape({
    selectCard: PropTypes.func,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    picture: PropTypes.string.isRequired,
    skills: skillsTypes
  })
};
