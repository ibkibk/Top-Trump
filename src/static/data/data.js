import craig from "../images/craig.png";
import liam from "../images/liam.png";

const cards = [
  {
    docId: "1",
    description:
      "Craig is Head of Technology for _nology. He loves rock climbing and solving problems. The only problem craig hasn't been able to solve is how to climb while leaving his ego on the ground.",
    name: "Craig Livings",
    picture: craig,
    role: "Trainer",
    skills: { loudness: 5, greed: 3, attitude: 8, sass: 7, mvp: 9 }
  },
  {
    docId: "2",
    description:
      "Liam is one of our great modern thinkers. He's always offering an opinion whether it is asked for or not. Ask Liam any question and he will give you any answer.",
    name: "Liam Farrell",
    picture: liam,
    role: "Trainer",
    skills: { loudness: 7, greed: 7, attitude: 5, sass: 5, mvp: 9 }
  }
];

export default cards;
