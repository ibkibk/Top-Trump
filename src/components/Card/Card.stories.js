import React from "react";
import Card from "./Card";
import craig from "../../static/images/craig.png";

export default {
  title: "Card"
};

const mockData = {
  description:
    "Craig is Head of Technology for _nology. He loves rock climbing and solving problems. The only problem craig hasn't been able to solve is how to climb while leaving his ego on the ground.",
  name: "Craig Livings",
  picture: craig,
  role: "Trainer",
  skills: { loudness: 5, greed: 3, attitude: 8, sass: 7, mvp: 9 }
};

export const card = () => <Card cardData={mockData} />;
