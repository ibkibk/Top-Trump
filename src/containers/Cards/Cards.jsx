import React, { Component } from "react";
import data from "../../static/data/data";
import styles from "./Cards.module.scss";
import Card from "../../components/Card";

class Cards extends Component {
  render() {
    return (
      <section className={styles.cards}>
        {/* map data array out to instances of card comp. 
            nologist referes to an individual data object & index 
            is the current position in the loop */}

        {data.map((nologist, index) => (
          <Card cardData={nologist} key={index} />
        ))}
      </section>
    );
  }
}

export default Cards;
