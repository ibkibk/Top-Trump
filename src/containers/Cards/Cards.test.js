import React from "react";
import { shallow } from "enzyme";
import Cards from "./Cards";
import data from "../../static/data/data";
import Card from "../../components/Card";

describe("Cards Tests", () => {
  let component;

  beforeEach(() => {
    component = shallow(<Cards />);
  });

  it("should render one instance of card component for each object of card data", () => {
    const numberCardsRendered = component.find(Card).length;
    const numberDataObjects = data.length;

    expect(numberCardsRendered).toEqual(numberDataObjects);
  });

  it("should match the snapshot", () => {
    expect(component).toMatchSnapshot();
  });
});
