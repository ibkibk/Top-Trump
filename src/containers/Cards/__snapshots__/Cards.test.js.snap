// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards Tests should match the snapshot 1`] = `
<section
  className="cards"
>
  <Card
    cardData={
      Object {
        "description": "Craig is Head of Technology for _nology. He loves rock climbing and solving problems. The only problem craig hasn't been able to solve is how to climb while leaving his ego on the ground.",
        "name": "Craig Livings",
        "picture": "craig.png",
        "role": "Trainer",
        "skills": Object {
          "attitude": 8,
          "greed": 3,
          "loudness": 5,
          "mvp": 9,
          "sass": 7,
        },
      }
    }
    key="0"
  />
  <Card
    cardData={
      Object {
        "description": "Liam is one of our great modern thinkers. He's always offering an opinion whether it is asked for or not. Ask Liam any question and he will give you any answer.",
        "name": "Liam Farrell",
        "picture": "liam.png",
        "role": "Trainer",
        "skills": Object {
          "attitude": 5,
          "greed": 7,
          "loudness": 7,
          "mvp": 9,
          "sass": 5,
        },
      }
    }
    key="1"
  />
</section>
`;
